airtime_mvc/application/common/TaskManager.php

```diff
@@ -0,0 +1,212 @@
+<?php
+
+/**
+ * Class TaskManager
+ */
+final class TaskManager {
+
+    /**
+     * @var array tasks to be run
+     */
+    protected $_taskList = [
+        AirtimeTask::UPGRADE,  // Always run the upgrade first
+        AirtimeTask::CELERY
+    ];
+
+    /**
+     * @var TaskManager singleton instance object
+     */
+    protected static $_instance;
+
+    /**
+     * @var int TASK_INTERVAL_SECONDS how often, in seconds, to run the TaskManager tasks,
+     *                                if they need to be run
+     */
+    const TASK_INTERVAL_SECONDS = 60;
+
+    /**
+     * @var $con PDO Propel connection object
+     */
+    private $_con;
+
+    /**
+     * Private constructor so class is uninstantiable
+     */
+    private function __construct() {
+    }
+
+    /**
+     * Get the singleton instance of this class
+     *
+     * @return TaskManager the TaskManager instance
+     */
+    public static function getInstance() {
+        if (!self::$_instance) {
+            self::$_instance = new TaskManager();
+        }
+        return self::$_instance;
+    }
+
+    /**
+     * Run all tasks that need to be run
+     */
+    public function runTasks() {
+        // If there is data in auth storage, this could be a user request
+        // so we should lock the TaskManager to avoid blocking
+        if ($this->_isUserSessionRequest()) return;
+        $this->_con = Propel::getConnection(CcPrefPeer::DATABASE_NAME);
+        $this->_con->beginTransaction();
+        try {
+            $lock = $this->_getLock();
+            if ($lock && microtime(true) < $lock['valstr'] + self::TASK_INTERVAL_SECONDS) return;
+            $this->_updateLock($lock);
+            $this->_con->commit();
+        } catch (Exception $e) {
+            // We get here if there are simultaneous requests trying to fetch the lock row
+            $this->_con->rollBack();
+            // Logging::info($e->getMessage()); // We actually get here a lot, so it's
+                                                // better to be silent here to avoid log bloat
+            return;
+        }
+        foreach ($this->_taskList as $task) {
+            $task = TaskFactory::getTask($task);
+            if ($task && $task->shouldBeRun()) $task->run();
+        }
+    }
+
+    /**
+     * Check if the current session is a user request
+     *
+     * @return bool
+     */
+    private function _isUserSessionRequest() {
+        $auth = Zend_Auth::getInstance();
+        $data = $auth->getStorage()->read();
+        return !empty($data);
+    }
+
+    /**
+     * Get the task_manager_lock from cc_pref with a row-level lock for atomicity
+     *
+     * @return array|bool an array containing the row values, or false on failure
```

Comment:
Should probably comment that this throws an exception if it is unable to immediately get the lock.
(Is that true?)
Also, did you test this?